rosbag2_compression/test/rosbag2_compression/test_sequential_compression_reader.cpp

```diff
@@ -37,49 +40,99 @@ class SequentialCompressionReaderTest : public Test
-    metadata.relative_file_paths = {"/path/to/storage"};
+    metadata.version = 4;
+    metadata.relative_file_paths = {
+      "bagfile_0.zstd",
+      "bagfile_1.zstd"
+    };
     metadata.topics_with_message_count.push_back({{topic_with_type_}, 1});
     metadata.compression_format = "zstd";
     metadata.compression_mode =
       rosbag2_compression::compression_mode_to_string(rosbag2_compression::CompressionMode::FILE);
     return metadata;
   }
 
-  std::unique_ptr<StrictMock<MockStorageFactory>> storage_factory_;
-  std::shared_ptr<NiceMock<MockStorage>> storage_;
-  std::shared_ptr<StrictMock<MockConverterFactory>> converter_factory_;
-  std::unique_ptr<NiceMock<MockMetadataIo>> metadata_io_;
+  void initialize_dummy_storage_files()
+  {
+    // Initialize some dummy files so that they can be found
+    rcpputils::fs::create_directories(tmp_dir_);
+    for (auto relative : metadata_.relative_file_paths) {
+      std::ofstream output((tmp_dir_ / relative).string());
+      output << "Fake storage data" << std::endl;
+    }
+  }
+
+  std::unique_ptr<rosbag2_compression::SequentialCompressionReader> create_reader()
+  {
+    auto decompressor = std::make_unique<NiceMock<MockDecompressor>>();
+    ON_CALL(*decompressor, decompress_uri).WillByDefault(
+      [](auto uri) {
+        auto path = rcpputils::fs::path(uri);
+        EXPECT_TRUE(path.exists());
+        return rcpputils::fs::remove_extension(path).string();
+      });
+    auto compression_factory = std::make_unique<NiceMock<MockCompressionFactory>>();
+    ON_CALL(*compression_factory, create_decompressor(_))
+    .WillByDefault(Return(ByMove(std::move(decompressor))));
+    return std::make_unique<rosbag2_compression::SequentialCompressionReader>(
+      std::move(compression_factory),
+      std::move(storage_factory_),
+      converter_factory_,
+      std::move(metadata_io_));
+  }
+
+  std::unique_ptr<MockStorageFactory> storage_factory_;
+  std::shared_ptr<MockStorage> storage_;
+  std::shared_ptr<MockConverterFactory> converter_factory_;
+  std::unique_ptr<MockMetadataIo> metadata_io_;
   std::unique_ptr<rosbag2_cpp::Reader> reader_;
   std::string storage_serialization_format_;
   rosbag2_storage::TopicMetadata topic_with_type_;
+  const std::string bag_name_ = "SequentialCompressionReaderTest";
+  rcpputils::fs::path tmp_dir_;
+  rosbag2_storage::StorageOptions storage_options_;
+  rosbag2_storage::BagMetadata metadata_;
+  rosbag2_cpp::ConverterOptions converter_options_;
 };
 
 TEST_F(SequentialCompressionReaderTest, open_throws_if_unsupported_compressor)
 {
-  rosbag2_storage::BagMetadata metadata = construct_default_bag_metadata();
-  metadata.compression_format = "bad_format";
-  EXPECT_CALL(*metadata_io_, read_metadata(_)).WillRepeatedly(Return(metadata));
-  EXPECT_CALL(*metadata_io_, metadata_file_exists(_)).WillRepeatedly(Return(true));
+  metadata_.compression_format = "bad_format";
+  EXPECT_CALL(*metadata_io_, read_metadata(_)).Times(AtLeast(1));
```

Comment:
Is `AtLeast(1)` what we want instead of `Times(1)`?
Is it possible that `read_metadata` is called more than 1x and that it is legal that it was?

Reply:
It ends up being exactly once, but I was translating the semantics of the existing test. My refactoring took the return value up to common code, but the `EXPECT` is specific to the test. I think I'd rather leave the logic unchanged here since I'm intentionally trying to leave all existing tests functionally unchanged